app/src/test/java/com/realityexpander/tasky/presentation/login_screen/LoginViewModelTest.kt

```diff
@@ -0,0 +1,188 @@
+        // ARRANGE
+        loginViewModel = LoginViewModel(
+            authRepository = authRepository,
+            validateEmail = validateEmail,
+            validatePassword = validatePassword,
+            savedStateHandle = SavedStateHandle()
+        )
+
+        // ACT
+        coroutineRule.dispatcher.scheduler.advanceUntilIdle()
+//        coroutineTestRule.testDispatcher.scheduler.advanceUntilIdle()
+
+        runTest {
+
+            loginViewModel.loginState.test {
+
+                // attempt to set state for password visibility. (is this the right way to do this?)
+                loginViewModel.sendEvent(LoginEvent.SetPasswordVisibility(true))
+                // https://developer.android.com/kotlin/flow/test
+                // https://medium.com/google-developer-experts/unit-testing-kotlin-flow-76ea5f4282c5
+
+                // ASSERT
+                var state = awaitItem() // initial state
+                println("isPasswordVisible: ${state.isPasswordVisible}")
+                assertThat(state.isPasswordVisible).isFalse()
+
+                // Await the "SetPasswordVisibility(true)" event
+                state = awaitItem()
+                assertThat(state.isPasswordVisible).isTrue() // why no worky?
+
+
+                ensureAllEventsConsumed()
+                cancelAndIgnoreRemainingEvents()
+                assertThat(loginViewModel.loginState.value.isPasswordVisible).isEqualTo(true)
+            }
+
+        }
+    }
+
+    @Test // broken
+    fun `LoginViewModel contains email and password and can set password visibility - ALTERNATE`()  {
+        // ARRANGE
+        loginViewModel = LoginViewModel(
+            authRepository = authRepository,
+            validateEmail = validateEmail,
+            validatePassword = validatePassword,
+            savedStateHandle = SavedStateHandle()
+        )
+
+        // ACT
+        coroutineRule.dispatcher.scheduler.advanceUntilIdle()
+//        coroutineTestRule.testDispatcher.scheduler.advanceUntilIdle()
+
+        runBlocking {
+            val job = launch {
+                loginViewModel.loginState.test {
+
+                    // ASSERT
+                    var state = awaitItem() // initial state
+                    println("isPasswordVisible: ${state.isPasswordVisible}")
+                    assertThat(state.isPasswordVisible).isFalse()
+
+                    // Await the "SetPasswordVisibility(true)" event
+                    state = awaitItem()
+                    assertThat(state.isPasswordVisible).isTrue() // why no worky?
+
+                    ensureAllEventsConsumed()
+                    cancelAndIgnoreRemainingEvents()
+                    assertThat(loginViewModel.loginState.value.isPasswordVisible).isEqualTo(true)
+                }
+            }
+
+            // attempt to set state for password visibility. (is this the right way to do this?)
+            loginViewModel.sendEvent(LoginEvent.SetPasswordVisibility(true))
```

Comment:
Everything that triggers state changes should happen inside the `test {}` block (I did this wrong in the past as well).
And yes, triggering the state change like that is the right way.

Reply:
Dang it, this one slipped past the goalie. I was using an event "Toggle" before I chose "Set", and that little exclamation mark caused me a lot of trouble!

Its working now.